### PR TITLE
Support 16 head disks, F9 as Suspend, and add enhanced diag disk and fixed install disk

### PR DESCRIPTION
**Pull request by arnoldrobbins**
Add enhanced diag and fixed install disk.
Support F9 as SUSPEND key for multiple terminals.

Having these changes will let uses create and install on big disks "out of the box". The F9 changes are per Jesse B, to support multiple login consoles as described here: https://stason.org/TULARC/pc/3b1-faq/4-4-How-do-I-get-multiple-login-windows.html.

### CODE REVIEW

**Comment by agentbooth** [comment]
Hey @arnoldrobbins -- quick note, if you fix the heads at 16 in wd2010_init(), the README still says to create an hd.img of 1024cyl x 8head x 17sec/tk.  The emulator will then detect this as a 512cyl x 16head x 17sec/tk image.  Which I assume will cause trouble if one attempts to format this as the Miniscribe 64MB drive. If one used the enhanced diag and entered 512, 16, 17 for the params, it would work. But that would be a little more advanced for a new user.

**Comment by arnoldrobbins** [comment]
Thanks for pointing this out.  @philpem This is another thing that should probably be configurable, instead of being hard coded.

I guess I can adjust the README some more. Personally, I think things should move to suggesting using the enhanced diag disk and formatting as big a disk as possible; that way there's lots of room left for playing. Let me know what I should do to get this PR accepted.

**Comment by agentbooth** [comment]
I didn't realize the format of the disk images -- having that geometry is very helpful :+1:!

**Comment by philpem** [comment]
I'm not certain you need the enhanced diags for 16-head/2-drive - the 3.51 diags seem to work, but the geometry has to be entered manually.

It's probably worth using the ICUS Enhanced Diags though.

**Comment by arnoldrobbins** [comment]
@philpem I agree - the regular diag disk does the trick. I need to still do some work in the README.md.  I have just pushed the tool I wrote to "prelabel" the file, and that closes the loop on reading the disk geometry from the file itself.  This PR is "almost" read to go now.

**Comment by arnoldrobbins** [comment]
@philpem I've revised `README.md`. I think this is ready to go now. Thanks!

File: README.md [hunk 1]
```diff
@@ -88,7 +90,7 @@ Phil Pemberton -- <philpem@philpem.me.uk>
 
   * [AT&T 3B1 Information](http://unixpc.taronga.com) -- the "Taronga archive".
     * Includes the STORE, comp.sources.3b1, XINU and a very easy to read HTML version of the 3B1 FAQ.
-    * Also includes (under "Kernel Related") tools to build an Enhanced Diagnostics disk which allows formatting hard drives with more than 8 heads or 1024 cylinders.  (We have provided a bootable disk in the `enhanced-diag` directory.)
+    * Also includes (under "Kernel Related") tools to build an Enhanced Diagnostics disk which allows formatting hard drives with more than 8 heads or 1024 cylinders.
```
Comment (agentbooth) [inline README.md]: Pretty sure Standard Diagnostics can format > 1024 cylinders -- just need Enhanced Diag for > 8 heads.

File: README.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ Phil Pemberton -- <philpem@philpem.me.uk>
     * Heads fixed at 16.
     * Sectors per track fixed at 17.
     * Fixed 512 bytes per sector.
```
Comment (agentbooth) [inline README.md:24]: Heads is now defaulted back to 8 (not 16).

File: src/main.c [hunk 3]
```diff
@@ -50,7 +50,7 @@ static int load_hd()
 		state.hdc_disc0 = NULL;
 		return (0);
 	}else{
-		wd2010_init(&state.hdc_ctx, state.hdc_disc0, 512, 17, 16);
+		wd2010_init(&state.hdc_ctx, state.hdc_disc0, 512, 16, 8);
```
Comment (agentbooth) [inline src/main.c]: Should this be 17 spt instead of 16?

**Comment by arnoldrobbins** [comment]
@shadyjesse Thanks for the comments. I have brought README.md into sync with them and with the code.

@philpem Any more comments?  I think this is ready for squash and merge...

Thanks to both of you!

**Comment by philpem** [comment]
That's a hell of a PR to review...

I'd have said do the second drive stuff as a separate PR for ease of review, and comment that it depends on this one... generally speaking the shorter the PR the easier it is to review and the faster it'll get merged.

Are you 100% done with this PR now, @arnoldrobbins ?

File: README.md [hunk 4]
```diff
@@ -50,43 +51,62 @@ Phil Pemberton -- <philpem@philpem.me.uk>
   - Download the 3B1 ROMs from Bitsavers: [link](http://bitsavers.org/pdf/att/3b1/firmware/3b1_roms.zip)
   - Download the 3B1 Foundation disk set from Bitsavers: [here](http://bitsavers.org/bits/ATT/unixPC/system_software_3.51/)
     * The disk images on unixpc.org don't work: the boot track is missing.
+    * Use the replacement version of the `08_Foundation_Set_Ver_3.51.IMD` image we have provided in the `os-install` directory instead of the one in the Foundation set.
```
Comment (philpem) [inline README.md]: Is this correct? We shouldn't be including code copyrighted to others in the repo. A shell script to take the image from Bitsavers and patch it (with MD5sum or SHA1sum check before and after) would be AOK.

File: README.md [hunk 5]
```diff
@@ -50,43 +51,62 @@ Phil Pemberton -- <philpem@philpem.me.uk>
...
   - Unzip the ROMs ZIP file and put the ROMs in a directory called `roms`:
     * Rename `14C 72-00616.bin` to `14c.bin`.
     * Rename `15C 72-00617.bin` to `15c.bin`.
   - Create a hard drive image file:
-    * `dd if=/dev/zero of=hd.img bs=512 count=$(expr 16 \* 8 \* 1024)`
-    * This creates a "Miniscribe 64MB" (CHS 1024:8:16, 512 bytes per sector).
-    * Note that you need the Enhanced Diagnostics disk to format 16-head hard drives.
-  - Install the operating system
+    * Use the `makehdimg` program supplied in the `tools` directory to create an initial `hd.img` file with the number of cylinders, heads and sectors per track that you want.  Limits: 1400 cylinders, 16 heads, 17 sectors per track.
```
Comment (philpem) [inline README.md:59]: I see Makehdimg source code, but no makefile or build instructions for it.

File: README.md [hunk 6]
```diff
@@ -50,43 +51,62 @@ Phil Pemberton -- <philpem@philpem.me.uk>
...
   * Alt-F12 -- Exit
 
 
 # Useful links
 
-  * [AT&T 3B1 Information](unixpc.taronga.com) -- the "Taronga archive".
+  * [AT&T 3B1 Information](http://unixpc.taronga.com) -- the "Taronga archive".
     * Includes the STORE, comp.sources.3b1, XINU and a very easy to read HTML version of the 3B1 FAQ.
-    * Also includes (under "Kernel Related") tools to build an Enhanced Diagnostics disk which allows formatting hard drives with more than 8 heads or 1024 cylinders.
+    * Also includes (under "Kernel Related") tools to build an Enhanced Diagnostics disk which allows formatting hard drives with more than 8 heads.
```
Comment (philpem) [inline README.md]: Are we sure about this? I've successfully formatted 1024x16 images with the 3.51 Diags disk.

File: os-install/README.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Fixed Installation Disk
+
+`08_Foundation_Set_Ver_3.51_no_phinit.IMD.gz` is a replacement for
```
Comment (philpem) [inline os-install/README.md]: Again, I'm not going to merge this if it or the branch contains code which is copyrighted to someone other than the authors, unless permission is granted by the current copyright holder.

File: src/main.c [hunk 8]
```diff
@@ -43,15 +43,27 @@ static int load_fd()
...
 static int load_hd()
 {
+	int ret = 0;
 
 	state.hdc_disc0 = fopen("hd.img", "r+b");
 	if (!state.hdc_disc0){
 		fprintf(stderr, "ERROR loading disc image 'hd.img'.\n");
 		state.hdc_disc0 = NULL;
 		return (0);
 	}else{
```
Comment (philpem) [inline src/main.c]: Nitpick alert: `} else {`

File: src/main.c [hunk 9]
```diff
@@ -43,15 +43,27 @@ static int load_fd()
...
+		printf("Disc image 'hd.img' loaded.\n");
+		ret = 1;
+	}
+
+	state.hdc_disc1 = fopen("hd2.img", "r+b");
+	if (!state.hdc_disc1){
+		fprintf(stderr, "ERROR loading disc image 'hd2.img'.\n");
+		state.hdc_disc1 = NULL;
+		return ret;
+	}else{
```
Comment (philpem) [inline src/main.c]: `} else {`

**Comment by arnoldrobbins** [comment]
@philpem I am done with this. I won't touch it further until you either ask me to make a change or you merge it.  I will try to do shorter PRs in the future.

Would it be useful for me to go through it and add explanatory comments / annotations?  Much thanks.

**Comment by philpem** [comment]
@arnoldrobbins I've reviewed it and left some comments. Just some minor code style fixes and some issues with the README.
Thanks.

File: README.md (hunk 5 above)
Comment (arnoldrobbins) [reply README.md:59]: Makefile supplied.

File: src/main.c (hunk 8 above)
Comment (arnoldrobbins) [reply src/main.c]: Fixed.

**Comment by arnoldrobbins** [comment]
@philpem I think I've addressed all your comments. Much thanks!

**Comment by philpem** [comment]
Sorry for the delay, I've been busy with Christmas and New Year stuff. Thanks for the PR, and happy new year!

**Pull request merged**